xiilib/django/charm.py

```diff
@@ -0,0 +1,87 @@
+#!/usr/bin/env python3
+
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Django Charm service."""
+import itertools
+import logging
+import pathlib
+import typing
+
+import ops
+
+# pydantic is causing this no-name-in-module problem
+from pydantic import BaseModel, Extra, Field, ValidationError  # pylint: disable=no-name-in-module
+
+from xiilib._gunicorn.charm import GunicornBase
+from xiilib.exceptions import CharmConfigInvalidError
+
+logger = logging.getLogger(__name__)
+
+
+class DjangoConfig(BaseModel, extra=Extra.allow):  # pylint: disable=too-few-public-methods
+    """Represent Django builtin configuration values.
+
+    Attrs:
+        debug: whether Django debug mode is enabled.
+        secret_key: a secret key that will be used for security related needs by your
+            Django application.
+        allowed_hosts: a list of host/domain names that this Django site can serve.
+    """
+
+    debug: bool | None = Field(None)
+    secret_key: str | None = Field(None, min_length=1)
+    allowed_hosts: list[str]
+
+
+class Charm(GunicornBase):  # pylint: disable=too-many-instance-attributes
+    """Django Charm service."""
+
+    def __init__(self, framework: ops.Framework) -> None:
+        """Initialize the Django charm.
+
+        Args:
+            framework: operator framework.
+        """
+        super().__init__(framework=framework, wsgi_framework="django")
+        self.framework.observe(self.on.django_app_pebble_ready, self._on_django_app_pebble_ready)
+
+    def get_wsgi_config(self) -> BaseModel:
+        """Return Django framework related configurations.
+
+        Returns:
+             Flask framework related configurations.
+
+        Raises:
+            CharmConfigInvalidError: if charm config is not valid.
+        """
+        django_config: dict[str, typing.Any] = {
+            "debug": self.config.get("django-debug"),
+            "secret_key": self.config.get("django-secret-key"),
```

Comment:
These 2 config keys are common between flask and django... I'm wondering if they shouldn't be at gunicorn level.
`debug` I'm pretty sure it'll be available in any framework / language
Less sure about secret_key

Reply:
Yes, I think we can make the decision if we have more samples. It's difficult to know if they will converge or diverge with only two samples.

Reply:
Ok, let's postpone the decision for later and keep it as it is